src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,228 @@
+        {
+            if (_lastEntryFound)
+                return false;
+
+            RawFindData findData = default;
+            do
+            {
+                FindNextFile();
+                if (!_lastEntryFound && _info != null)
+                {
+                    // If needed, stash any subdirectories to process later
+                    if (_recursive && (_info->FileAttributes & FileAttributes.Directory) != 0
+                        && !PathHelpers.IsDotOrDotDot(_info->FileName))
+                    {
+                        string subDirectory = PathHelpers.CombineNoChecks(_currentPath, _info->FileName);
+                        _pending.Enqueue((CreateDirectoryHandle(subDirectory), subDirectory));
+                    }
+
+                    findData = new RawFindData(_info, _currentPath, _originalFullPath, _originalUserPath);
+                }
+            } while (!_lastEntryFound && !_predicate(ref findData, _state));
+
+            return !_lastEntryFound;
+        }
+
+        private unsafe void FindNextFile()
+        {
+            Interop.NtDll.FILE_FULL_DIR_INFORMATION* info = _info;
+            if (info != null && info->NextEntryOffset != 0)
+            {
+                // We're already in a buffer and have another entry
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)((byte*)info + info->NextEntryOffset);
+                return;
+            }
+
+            // We need more data
+            if (GetData())
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)_pinnedBuffer.AddrOfPinnedObject();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void NoMoreFiles()
+        {
+            _info = null;
+            if (_pending == null || _pending.Count == 0)
+            {
+                _lastEntryFound = true;
+            }
+            else
+            {
+                // Grab the next directory to parse
+                (IntPtr Handle, string Path) next = _pending.Dequeue();
+                Interop.Kernel32.CloseHandle(_directoryHandle);
+                _directoryHandle = next.Handle;
+                _currentPath = next.Path;
+                FindNextFile();
+            }
+        }
+
+        public void Reset()
+        {
+            throw new NotSupportedException();
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected void Dispose(bool disposing)
+        {
+            byte[] buffer = Interlocked.Exchange(ref _buffer, null);
+            if (buffer != null)
+            {
+                Interop.Kernel32.CloseHandle(_directoryHandle);
+                if (_recursive && _pending != null)
+                {
+                    while (_pending.Count > 0)
+                        Interop.Kernel32.CloseHandle(_pending.Dequeue().Handle);
+                }
+
+                _pinnedBuffer.Free();
```

Comment:
`GCHandle.Free` throws on null handle. Check for IsAllocated before calling Free?

Reply:
Ok